frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -3217,6 +3217,45 @@ def tree_add(self, resulting_type: ConversionType, xs: List[Conversion]) -> Conv
 
         return xs[0]
 
+    def truncate_bit_pattern(self, x: Conversion, lsbs_to_remove: int) -> Conversion:
+        if x.is_clear:
+            highlights = {
+                x.origin: "operand is clear",
+                self.converting: "but clear truncate bit pattern is not supported",
+            }
+            self.error(highlights)
+
+        assert x.bit_width > lsbs_to_remove
+
+        resulting_bit_width = x.bit_width
```

Comment:
That is the execution semantics. `truncate_bit_pattern` is for setting the last `n` bits to `0` in a fast way. Because of this, when you do:
```python
for x in range(32):
    print(f"truncate({x=}, lsbs_to_remove=2) == {fhe.truncate_bit_pattern(x, lsbs_to_remove=2)}")
```
you get:
```
truncate(x=0, lsbs_to_remove=2) == 0
truncate(x=1, lsbs_to_remove=2) == 0
truncate(x=2, lsbs_to_remove=2) == 0
truncate(x=3, lsbs_to_remove=2) == 0
truncate(x=4, lsbs_to_remove=2) == 4
truncate(x=5, lsbs_to_remove=2) == 4
truncate(x=6, lsbs_to_remove=2) == 4
truncate(x=7, lsbs_to_remove=2) == 4
truncate(x=8, lsbs_to_remove=2) == 8
truncate(x=9, lsbs_to_remove=2) == 8
truncate(x=10, lsbs_to_remove=2) == 8
truncate(x=11, lsbs_to_remove=2) == 8
truncate(x=12, lsbs_to_remove=2) == 12
truncate(x=13, lsbs_to_remove=2) == 12
truncate(x=14, lsbs_to_remove=2) == 12
truncate(x=15, lsbs_to_remove=2) == 12
truncate(x=16, lsbs_to_remove=2) == 16
truncate(x=17, lsbs_to_remove=2) == 16
truncate(x=18, lsbs_to_remove=2) == 16
truncate(x=19, lsbs_to_remove=2) == 16
truncate(x=20, lsbs_to_remove=2) == 20
truncate(x=21, lsbs_to_remove=2) == 20
truncate(x=22, lsbs_to_remove=2) == 20
truncate(x=23, lsbs_to_remove=2) == 20
truncate(x=24, lsbs_to_remove=2) == 24
truncate(x=25, lsbs_to_remove=2) == 24
truncate(x=26, lsbs_to_remove=2) == 24
truncate(x=27, lsbs_to_remove=2) == 24
truncate(x=28, lsbs_to_remove=2) == 28
truncate(x=29, lsbs_to_remove=2) == 28
truncate(x=30, lsbs_to_remove=2) == 28
truncate(x=31, lsbs_to_remove=2) == 28
```
and if you apply a TLU:
```python
for x in range(32):
    print(f"truncate({x=}, lsbs_to_remove=2) ** 2 == {fhe.truncate_bit_pattern(x, lsbs_to_remove=2) ** 2}")
```
you get:
```
truncate(x=0, lsbs_to_remove=2) ** 2 == 0
truncate(x=1, lsbs_to_remove=2) ** 2 == 0
truncate(x=2, lsbs_to_remove=2) ** 2 == 0
truncate(x=3, lsbs_to_remove=2) ** 2 == 0
truncate(x=4, lsbs_to_remove=2) ** 2 == 16
truncate(x=5, lsbs_to_remove=2) ** 2 == 16
truncate(x=6, lsbs_to_remove=2) ** 2 == 16
truncate(x=7, lsbs_to_remove=2) ** 2 == 16
truncate(x=8, lsbs_to_remove=2) ** 2 == 64
truncate(x=9, lsbs_to_remove=2) ** 2 == 64
truncate(x=10, lsbs_to_remove=2) ** 2 == 64
truncate(x=11, lsbs_to_remove=2) ** 2 == 64
truncate(x=12, lsbs_to_remove=2) ** 2 == 144
truncate(x=13, lsbs_to_remove=2) ** 2 == 144
truncate(x=14, lsbs_to_remove=2) ** 2 == 144
truncate(x=15, lsbs_to_remove=2) ** 2 == 144
truncate(x=16, lsbs_to_remove=2) ** 2 == 256
truncate(x=17, lsbs_to_remove=2) ** 2 == 256
truncate(x=18, lsbs_to_remove=2) ** 2 == 256
truncate(x=19, lsbs_to_remove=2) ** 2 == 256
truncate(x=20, lsbs_to_remove=2) ** 2 == 400
truncate(x=21, lsbs_to_remove=2) ** 2 == 400
truncate(x=22, lsbs_to_remove=2) ** 2 == 400
truncate(x=23, lsbs_to_remove=2) ** 2 == 400
truncate(x=24, lsbs_to_remove=2) ** 2 == 576
truncate(x=25, lsbs_to_remove=2) ** 2 == 576
truncate(x=26, lsbs_to_remove=2) ** 2 == 576
truncate(x=27, lsbs_to_remove=2) ** 2 == 576
truncate(x=28, lsbs_to_remove=2) ** 2 == 784
truncate(x=29, lsbs_to_remove=2) ** 2 == 784
truncate(x=30, lsbs_to_remove=2) ** 2 == 784
truncate(x=31, lsbs_to_remove=2) ** 2 == 784
```
If we were to reduce the precision, this is what would happen:
```python
for x in range(32):
    print(f"truncate({x=}, lsbs_to_remove=2) == {fhe.truncate_bit_pattern(x, lsbs_to_remove=2) // 4}")

print()

for x in range(32):
    print(f"truncate({x=}, lsbs_to_remove=2) ** 2 == {fhe.truncate_bit_pattern(x, lsbs_to_remove=2) ** 2 // 4}")
```
would result in:
```
truncate(x=0, lsbs_to_remove=2) == 0
truncate(x=1, lsbs_to_remove=2) == 0
truncate(x=2, lsbs_to_remove=2) == 0
truncate(x=3, lsbs_to_remove=2) == 0
truncate(x=4, lsbs_to_remove=2) == 1
truncate(x=5, lsbs_to_remove=2) == 1
truncate(x=6, lsbs_to_remove=2) == 1
truncate(x=7, lsbs_to_remove=2) == 1
truncate(x=8, lsbs_to_remove=2) == 2
truncate(x=9, lsbs_to_remove=2) == 2
truncate(x=10, lsbs_to_remove=2) == 2
truncate(x=11, lsbs_to_remove=2) == 2
truncate(x=12, lsbs_to_remove=2) == 3
truncate(x=13, lsbs_to_remove=2) == 3
truncate(x=14, lsbs_to_remove=2) == 3
truncate(x=15, lsbs_to_remove=2) == 3
truncate(x=16, lsbs_to_remove=2) == 4
truncate(x=17, lsbs_to_remove=2) == 4
truncate(x=18, lsbs_to_remove=2) == 4
truncate(x=19, lsbs_to_remove=2) == 4
truncate(x=20, lsbs_to_remove=2) == 5
truncate(x=21, lsbs_to_remove=2) == 5
truncate(x=22, lsbs_to_remove=2) == 5
truncate(x=23, lsbs_to_remove=2) == 5
truncate(x=24, lsbs_to_remove=2) == 6
truncate(x=25, lsbs_to_remove=2) == 6
truncate(x=26, lsbs_to_remove=2) == 6
truncate(x=27, lsbs_to_remove=2) == 6
truncate(x=28, lsbs_to_remove=2) == 7
truncate(x=29, lsbs_to_remove=2) == 7
truncate(x=30, lsbs_to_remove=2) == 7
truncate(x=31, lsbs_to_remove=2) == 7

truncate(x=0, lsbs_to_remove=2) ** 2 == 0
truncate(x=1, lsbs_to_remove=2) ** 2 == 0
truncate(x=2, lsbs_to_remove=2) ** 2 == 0
truncate(x=3, lsbs_to_remove=2) ** 2 == 0
truncate(x=4, lsbs_to_remove=2) ** 2 == 4
truncate(x=5, lsbs_to_remove=2) ** 2 == 4
truncate(x=6, lsbs_to_remove=2) ** 2 == 4
truncate(x=7, lsbs_to_remove=2) ** 2 == 4
truncate(x=8, lsbs_to_remove=2) ** 2 == 16
truncate(x=9, lsbs_to_remove=2) ** 2 == 16
truncate(x=10, lsbs_to_remove=2) ** 2 == 16
truncate(x=11, lsbs_to_remove=2) ** 2 == 16
truncate(x=12, lsbs_to_remove=2) ** 2 == 36
truncate(x=13, lsbs_to_remove=2) ** 2 == 36
truncate(x=14, lsbs_to_remove=2) ** 2 == 36
truncate(x=15, lsbs_to_remove=2) ** 2 == 36
truncate(x=16, lsbs_to_remove=2) ** 2 == 64
truncate(x=17, lsbs_to_remove=2) ** 2 == 64
truncate(x=18, lsbs_to_remove=2) ** 2 == 64
truncate(x=19, lsbs_to_remove=2) ** 2 == 64
truncate(x=20, lsbs_to_remove=2) ** 2 == 100
truncate(x=21, lsbs_to_remove=2) ** 2 == 100
truncate(x=22, lsbs_to_remove=2) ** 2 == 100
truncate(x=23, lsbs_to_remove=2) ** 2 == 100
truncate(x=24, lsbs_to_remove=2) ** 2 == 144
truncate(x=25, lsbs_to_remove=2) ** 2 == 144
truncate(x=26, lsbs_to_remove=2) ** 2 == 144
truncate(x=27, lsbs_to_remove=2) ** 2 == 144
truncate(x=28, lsbs_to_remove=2) ** 2 == 196
truncate(x=29, lsbs_to_remove=2) ** 2 == 196
truncate(x=30, lsbs_to_remove=2) ** 2 == 196
truncate(x=31, lsbs_to_remove=2) ** 2 == 196
```

And `31**2` is much closer to `784` than `196`. This is how the ML team wants to use the truncated values for example.